File: tests/mv-00/dst_00.java

import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

@SuppressWarnings("unchecked")
public class dst_00 extends ModelChecker
{
    static public final String VERSION = "2";

    public Map<String, Checker> dst_00_map_pmap;

    public boolean _jm_obj_0(Object val, Path path, Report rep)
    {
        if (json.objectSize(val) != 2)
        {
            return false;
        }
        Object pval;
        if (! ((pval = json.objectValue(val, "a")) != null))
        {
            return false;
        }
        boolean res = json.isInteger(pval) && json.asLong(pval) >= 0;
        if (! res)
        {
            return false;
        }
        if (! ((pval = json.objectValue(val, "c")) != null))
        {
            return false;
        }
        return json.isString(pval);
    }

    public boolean _jm_obj_1(Object val, Path path, Report rep)
    {
        if (json.objectSize(val) != 2)
        {
            return false;
        }
        Object pval;
        if (! ((pval = json.objectValue(val, "a")) != null))
        {
            return false;
        }
        boolean res = json.isInteger(pval) && json.asLong(pval) >= 0;
        if (! res)
        {
            return false;
        }
        if (! ((pval = json.objectValue(val, "b")) != null))
        {
            return false;
        }
        return json.isInteger(pval) && json.asLong(pval) >= 1;
    }

    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res = json.isObject(val);
        if (res)
        {
            if (json.objectHasProp(val, "b"))
            {
                res = _jm_obj_1(val, null, null);
            }
            else
            {
                res = json.objectHasProp(val, "c") && _jm_obj_0(val, null, null);
            }
        }
        return res;
    }


    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            dst_00_map_pmap = new HashMap<String, Checker>();
            dst_00_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            dst_00_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return dst_00_map_pmap.get(name);
    }

    public Set<String> models()
    {
        return dst_00_map_pmap.keySet();
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new dst_00();
        Main.main("dst_00", checker, VERSION, args);
    }
}

File: tests/mv-00/dst_00.py
#! /bin/env python
import math
from typing import Callable
import re as re
from json_model.runtime import *
__version__ = "2"

def check_model_fun(name: str) -> CheckFun:
    """Return check function for JSON model name."""
    return check_model_map[name]

# entry point for generated checkers
def check_model(val: Jsonable, name: str = "", rep: Report = None) -> bool:
    """Check val validity against JSON Model name."""
    if name not in check_model_map:
        raise Exception(f"unexpected model name: {name}")
    checker = check_model_map[name]
    return checker(val, [], rep)

check_model_map: PropMap

def _jm_obj_0(val: Jsonable, path: Path, rep: Report) -> bool:
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("a", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, int) and not isinstance(pval, bool) and pval >= 0
    if not res:
        return False
    if not ((pval := val.get("c", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, str)

def _jm_obj_1(val: Jsonable, path: Path, rep: Report) -> bool:
    if len(val) != 2:
        return False
    pval: Jsonable
    if not ((pval := val.get("a", UNDEFINED)) != UNDEFINED):
        return False
    res: bool = isinstance(pval, int) and not isinstance(pval, bool) and pval >= 0
    if not res:
        return False
    if not ((pval := val.get("b", UNDEFINED)) != UNDEFINED):
        return False
    return isinstance(pval, int) and not isinstance(pval, bool) and pval >= 1

def json_model_1(val: Jsonable, path: Path, rep: Report) -> bool:
    res: bool = isinstance(val, dict)
    if res:
        if "b" in val:
            res = _jm_obj_1(val, None, None)
        else:
            res = "c" in val and _jm_obj_0(val, None, None)
    return res


# initialization guard
initialized: bool = False

# differed module initializations
def check_model_init():
    global initialized
    if not initialized:
        initialized = True
        global check_model_map
        check_model_map = {
            "": json_model_1,
        }

# differed module cleanup
def check_model_free():
    global initialized
    if initialized:
        initialized = False

if __name__ == "__main__":
    check_model_init()
    main(check_model_fun, check_model_map, __version__)
    check_model_free()
